### PR TITLE
Add API to create PKCS#12

### PR DESCRIPTION
**Pull request by gjcairo**
## Motivation
It would be handy to provide an API to create PKCS#12 files from a list of `NIOSSLCertificates` and a `NIOSSLPrivateKey`.
This would be particularly useful when dealing with Network.framework/NIOTransportServices/Security.framework, which use `SecIdentity`s for SSL. Two particular use cases are https://github.com/apple/swift-nio-ssl/issues/484#issuecomment-2410375188 and `grpc-swift-nio-transport`, which would use this API for testing the NIOTS transport implementation.

## Modifications
This PR adds a static method to `NIOSSLPKCS12Bundle` that creates a PKCS#12 file from the given array of certificates + private key, and returns it as an array of bytes.

## Result
PKCS#12 files can be created using NIOSSL.

### CODE REVIEW

File: Sources/NIOSSL/SSLPKCS12Bundle.swift [hunk 1]
```diff
@@ -158,6 +158,86 @@ public struct NIOSSLPKCS12Bundle: Hashable {
...
+    ///   - certificates: An array of certificates to include in this PKCS#12. Must contain at least one certificate.
+    ///   - privateKey: The private key associated to the first certificate in `certificates`.
+    ///   - passphrase: The password with which to protect this PKCS#12 file.
+    ///   - name: The name to give this PKCS#12 file.
+    /// - Returns: An array of bytes making up the PKCS#12 file.
+    public static func makePKCS12<Bytes: Collection>(
+        certificates: [NIOSSLCertificate],
+        privateKey: NIOSSLPrivateKey,
+        passphrase: Bytes,
+        name: Bytes
```
Comment (Lukasa) [inline Sources/NIOSSL/SSLPKCS12Bundle.swift:179]: We don't need to provide a name, and I'd argue we shouldn't bother. If we're going to enable this, it should be optional, and needs a different generic type parameter to the passphrase.

File: Sources/NIOSSL/SSLPKCS12Bundle.swift [hunk 2]
```diff
@@ -158,6 +158,86 @@ public struct NIOSSLPKCS12Bundle: Hashable {
...
+    ) throws -> [UInt8] where Bytes.Element == UInt8 {
```
Comment (Lukasa) [inline Sources/NIOSSL/SSLPKCS12Bundle.swift:180]: I think this API does too much. Should we split it out? We can have a memberwise constructor on the PKCS12, and then a serialize function.

File: Sources/NIOSSL/SSLPKCS12Bundle.swift [hunk 3]
```diff
@@ -158,6 +158,86 @@ public struct NIOSSLPKCS12Bundle: Hashable {
...
+        guard let mainCertificate = certificates.first else {
+            preconditionFailure("At least one certificate must be provided")
+        }
+
+        let certificateChainStack = CNIOBoringSSL_sk_X509_new(nil)
+        for additionalCertificate in certificates.dropFirst() {
+            let result = additionalCertificate.withUnsafeMutableX509Pointer { certificate in
+                CNIOBoringSSL_sk_X509_push(certificateChainStack, certificate)
```
Comment (Lukasa) [inline Sources/NIOSSL/SSLPKCS12Bundle.swift:188]: This gets the refcount wrong. We need to increase the refcount of the certificate pointer here, using `X509_up_ref`

File: Sources/NIOSSL/SSLPKCS12Bundle.swift [hunk 4]
```diff
@@ -158,6 +158,86 @@ public struct NIOSSLPKCS12Bundle: Hashable {
...
+            }
+            if result == 0 {
+                fatalError("Failed to add certificate to chain")
+            }
+        }
+
+        let pkcs12 = try passphrase.withSecureCString { passphrase in
```
Comment (Lukasa) [inline Sources/NIOSSL/SSLPKCS12Bundle.swift:195]: There's no matching `PKCS12_free` here.

File: Sources/NIOSSL/SSLPKCS12Bundle.swift (hunk 2 above)
Comment (gjcairo) [reply Sources/NIOSSL/SSLPKCS12Bundle.swift:180]: Do you mean having an `init(certificateChain: [NIOSSLCertificate], privateKey: NIOSSLPrivateKey)` on `NIOSSLPKCS12Bundle`, and then a `serialize(passphrase: Bytes) -> [UInt8]`?

File: Sources/NIOSSL/SSLPKCS12Bundle.swift (hunk 1 above)
Comment (gjcairo) [reply Sources/NIOSSL/SSLPKCS12Bundle.swift:179]: I'll just get rid of it, don't think it serves any purpose really.

File: Sources/NIOSSL/SSLPKCS12Bundle.swift (hunk 3 above)
Comment (gjcairo) [reply Sources/NIOSSL/SSLPKCS12Bundle.swift:188]: Ah, the joys of manual memory management. 
Should we decrease the refcount after we've created the PKCS12?

File: Sources/NIOSSL/SSLPKCS12Bundle.swift (hunk 2 above)
Comment (Lukasa) [reply Sources/NIOSSL/SSLPKCS12Bundle.swift:180]: Yeah.

File: Sources/NIOSSL/SSLPKCS12Bundle.swift (hunk 3 above)
Comment (Lukasa) [reply Sources/NIOSSL/SSLPKCS12Bundle.swift:188]: Yes, we're also missing the free of the `sk_509`. Done correctly, the free call is `sk_X509_pop_free` with the second argument as `X509_free`.

File: Sources/NIOSSL/SSLPKCS12Bundle.swift [hunk 5]
```diff
@@ -158,6 +158,101 @@ public struct NIOSSLPKCS12Bundle: Hashable {
...
+            let result = additionalCertificate.withUnsafeMutableX509Pointer { certificate in
+                CNIOBoringSSL_X509_up_ref(certificate)
+                return CNIOBoringSSL_sk_X509_push(certificateChainStack, certificate)
+            }
+            if result == 0 {
+                fatalError("Failed to add certificate to chain")
+            }
+        }
+
+        defer {
```
Comment (Lukasa) [inline Sources/NIOSSL/SSLPKCS12Bundle.swift:205]: Nit: as a matter of coding style it's a bit nicer to move this immediately after the `sk_X509_new`.

File: Tests/NIOSSLTests/SSLPKCS12BundleTest.swift [hunk 6]
```diff
@@ -527,4 +527,28 @@ class SSLPKCS12BundleTest: XCTestCase {
...
+
+        // Create a PKCS#12...
+        let bundle = NIOSSLPKCS12Bundle(
+            certificateChain: certificates,
+            privateKey: privateKey
+        )
+        let pkcs12 = try bundle.serialize(passphrase: "thisisagreatpassword".utf8)
+
+        // And then decode it into a NIOSSLPKCS12Bundle
+        let decoded = try NIOSSLPKCS12Bundle(buffer: pkcs12, passphrase: "thisisagreatpassword".utf8)
```
Comment (Lukasa) [inline Tests/NIOSSLTests/SSLPKCS12BundleTest.swift:548]: We should probably try some non-matching passphrases.

**Review by Lukasa — APPROVED** [review]

**Pull request merged**